include/zephyr/drivers/video.h

```diff
@@ -753,104 +753,185 @@ struct video_buffer *video_buffer_alloc(size_t size);
-#define video_fourcc(a, b, c, d)                                                                   \
-	((uint32_t)(a) | ((uint32_t)(b) << 8) | ((uint32_t)(c) << 16) | ((uint32_t)(d) << 24))
-
 /**
  * @defgroup video_pixel_formats Video pixel formats
+ * The @c | characters separate the pixels, and spaces separate the bytes.
+ * The uppercase letter represents the most significant bit.
+ * The lowercase letters represent the rest of the bits.
  * @{
  */
 
 /**
- * @name Bayer formats
+ * @brief Four-character-code uniquely identifying the pixel format
+ */
+#define VIDEO_FOURCC(a, b, c, d)                                                                   \
+	((uint32_t)(a) | ((uint32_t)(b) << 8) | ((uint32_t)(c) << 16) | ((uint32_t)(d) << 24))
+
+/**
+ * @brief Convert a four-character-string to a four-character-code
+ *
+ * Convert a string literal or variable into a four-character-code
+ * as defined by @ref VIDEO_FOURCC.
+ *
+ * @param str String to be converted
+ * @return Four-character-code.
+ */
+#define VIDEO_FOURCC_FROM_STR(str) VIDEO_FOURCC((str)[0], (str)[1], (str)[2], (str)[3])
+
+/**
+ * @name Bayer formats (R, G, B channels).
  * @{
  */
 
-/** BGGR8 pixel format */
-#define VIDEO_PIX_FMT_BGGR8 video_fourcc('B', 'G', 'G', 'R') /*  8  BGBG.. GRGR.. */
-/** GBRG8 pixel format */
-#define VIDEO_PIX_FMT_GBRG8 video_fourcc('G', 'B', 'R', 'G') /*  8  GBGB.. RGRG.. */
-/** GRBG8 pixel format */
-#define VIDEO_PIX_FMT_GRBG8 video_fourcc('G', 'R', 'B', 'G') /*  8  GRGR.. BGBG.. */
-/** RGGB8 pixel format */
-#define VIDEO_PIX_FMT_RGGB8 video_fourcc('R', 'G', 'G', 'B') /*  8  RGRG.. GBGB.. */
+/**
+ * @verbatim
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ...
+ * | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_BGGR8 VIDEO_FOURCC('B', 'A', '8', '1')
+
+/**
+ * @verbatim
+ * | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | ...
+ * | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_GBRG8 VIDEO_FOURCC('G', 'B', 'R', 'G')
+
+/**
+ * @verbatim
+ * | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | ...
+ * | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_GRBG8 VIDEO_FOURCC('G', 'R', 'B', 'G')
+
+/**
+ * @verbatim
+ * | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | ...
+ * | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | ...
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_RGGB8 VIDEO_FOURCC('R', 'G', 'G', 'B')
 
 /**
  * @}
  */
 
 /**
  * @name RGB formats
+ * Per-color (R, G, B) channels.
  * @{
  */
 
-/** RGB565 pixel format */
-#define VIDEO_PIX_FMT_RGB565 video_fourcc('R', 'G', 'B', 'P') /* 16  RGB-5-6-5 */
+/**
+ * 5 red bits [4:0], 6 green bits [10:5], 5 blue bits [15:11].
+ * In big endian format over two bytes.
```

Comment:
Why this format is described in BE while the RGB565 is described in LE, why we don't describe both in either LE or BE ?

Reply:
Linux's [`V4L2_PIX_FMT_RGB565X`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/videodev2.h#n564) is the big endian version of [`V4L2_PIX_FMT_RGB565`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/videodev2.h#n560).

This text description documents the bits, then says this should be packed as little-endian, like in the text bit diagram.

This matches the C operations done to decode the format:
```c
uint16_t rgb = sys_be16_to_cpu(rgb565x); // Or sys_le16_to_cpu() for little-endian RGB565
uint8_t r = (rgb >> 11) << 3;
uint8_t g = (rgb >> 5) << 2;
uint8_t b = (rgb >> 0) << 3;
```

Before I figured out the reference material I used (LWN) was wrong, I documented the bits to be "red first", but assigned to the wrong bits to it ([4:0] instead of [15:11]). I adjusted this in the latest commit.

Do you mean you would rather have this description format for i.e. Little Endian?

```
 * bits [2:0] of green, bits [4:0] of blue, bits [4:0] of red, bits [5:3] of green.
```

Do you mean we should opt for RGB565LE and RGB565BE [like done in libav](https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavutil/pixfmt.h#L98-L106) (for instance)? This is different from Linux source but less ambiguous.

Reply:
Here I just mean the consistency. If we documented RGB565 in LE, so we should also talk about other formats in LE too, or vice versa. But this is just a small detail.

Reply:
It looks like insofar, `RGB565` (BE/LE) is the only format where a field spans across multiple bytes.
Other formats are all described by byte access with no big-endian/little-endian variant AFAIU:

- Bayer formats: 8-bit per pixel: no byte endianess issue
- `XRGB32` (and `XYUV` and `YUYV`): There is a `BGRX32`, just like there is a `BGR565`, but no `XRGB32-LE` (or `XRGB32X`):
  This is 8-bit per channel, so different endianess == different channel order

I avoided to add more in case it was considered as unnecessary.
Let me know if I should add i.e. "all little-endian counterparts" or "all of Linux" or "all the most common" etc. :)

I added a comment mentioning that a 16-bit integer had to be split, which does not happen for the other formats.

Reply:
It's ok. Thanks.